pkg/ipcache/cidr.go

```diff
@@ -64,6 +64,7 @@ func (ipc *IPCache) AllocateCIDRs(
 		id, isNew, err := ipc.allocate(p, lbls, oldNID)
 		if err != nil {
 			ipc.IdentityAllocator.ReleaseSlice(context.Background(), nil, usedIdentities)
+			ipc.Unlock()
```

Comment:
a naive question, but i guess there is a reason we can't just defer Unlock?

Reply:
@michi-covalent looks like the call to:
```
		ipc.UpsertGeneratedIdentities(newlyAllocatedIdentities, nil)
```

Tries to grab the lock too. This is kinda tricky code, but could dead lock I believe if its not organized like it is here.

Reply:
lock, unlock, and then lock again huh. struggle but i guess that's how it is ✅

Reply:
We _could_ defer `Unlock()` and also change the call into `ipc.Upsert()` below, maybe that code is simpler. But in a way it's also more complex, because depending on whether the caller handles the `Upsert` or not, the locking may cover that portion of the logic or not.